### ISSUE TITLE
Retraining Vgg from existing weights

### ISSUE THREAD

**Issue by saivineethkumar** [AlexeyAB/darknet#1061]
Hi @AlexeyAB 
I want to retrain the vgg model for classification of more objects, that is presently we can classify 1000 classes using imagenet, but i want to increase this to something like 1010. And I dont want to train the model from scratch, I want to use the existing weights and change the last layer to the classify 1010 classes.
I tried similar steps that you listed for cifar training but could not achieve success as cifar is trained from scratch but i want to train vgg from existing weights.

Could you tell me the exact steps to be followed and the changes that are to be made to the files to achieve this task.
Thank you!

**Comment by saivineethkumar** [comment]
@AlexeyAB  thanks for the reply

> And your labels should have indexies from 1000 to 1009

where can i find the label list file to update? can u specify the location.

And when I said I dont want to train from scratch, I didnt mean no training at all. what i am interested in is that i want to use the existing `vgg-16.weights` file as a starting point to train the model for new classes using some 200 images per each class so that the weights in the last layers will be altered a bit when trained, to get decent enough accuracy for the new classes without decreasing the accuracy of the previous model much( i am aware that by doing this accuracy of the previous model will decrease).
So is it possible to train the model like this, where existing weights are used and altered a bit when trained on new image set.

thanks!  and sorry if some of these doubts seem silly. I am new to this and would really appreciate your help.

**Comment by AlexeyAB** [comment]
`imagenet.shortnames.list` https://github.com/AlexeyAB/darknet/blob/7f0079dec5da19072cfe64aee233ecacd8ef2ffd/cfg/imagenet.shortnames.list#L1000-L1009

`imagenet.labels.list` https://github.com/AlexeyAB/darknet/blob/7f0079dec5da19072cfe64aee233ecacd8ef2ffd/cfg/imagenet.labels.list#L1000-L1009

**Comment by saivineethkumar** [comment]
@AlexeyAB 

I updated the lists as you said and when I run the following command:
`./darknet classifier train cfg/test2.data cfg/test.cfg backup/test.weights` to train the model to make it learn the new classes I am getting the following error:

`Too many or too few labels: 0, /home/sai/darknettest/data/cifar/train/rose78.jpg` for all the training images. the label `rose` is there in the updated list, still this error shows up.

and at the end it shows this 
```
Saving weights to /home/darknettest/backup/test_14872.weights
Couldn't open file: /home/darknettest/backup/test_14872.weights
```
And I had updated the number of classes in the above test.cfg file too. 
Do you have any idea about this error? 

thank you!

**Comment by saivineethkumar** [comment]
@AlexeyAB 
I am able to resolve the above error and train the model. The model is predicting the new classes well but fail to retain its accuracy for the old classes ( it is not at all predicting the old classes correctly). 

**This is how i wanted to train my model:**
I want to expand the last layer from 1000 to 1005 neurons so that the weights on the old links will be retained and the new ones will be randomized. when I start training, these weights will be altered according to GD algorithm. I will keep an eye on number of epochs to train based on some threshold value of accuracy of prediction of old classes. Now after completing the training the accuracy of the old classes will not be below a threshold and for the new ones, it will be good enough.

So my question is: Is it possible to train my model like this? Can you suggest me a way to do training like this. I dont want to start training from scratch, I want to use the existing weights as starting point for training.
Thanks!

**Comment by AlexeyAB** [comment]
@saivineethkumar 

> I am able to resolve the above error and train the model. The model is predicting the new classes well but fail to retain its accuracy for the old classes ( it is not at all predicting the old classes correctly).

Delete this weights-file.

Try to use fine-tuning, just set `stopbackward=1` in the penultimate [convolutional]-layer in cfg-file, and train in the same way as you done.

**Comment by saivineethkumar** [comment]
My doubt is that, if we change the last layer neurons from 1000 to 1005 and we are importing the pretrained weight file which is trained for 1000 neurons, will it fit our new model ? if yes, will the other weights that are not loaded will be randomly assigned?

I changed the neurons from 1000 to 1005 and checked the accuracy of the already existing class, the result is that its not even detecting the class label correctly. Do you have any idea about why this happens like this?

**Comment by AlexeyAB** [comment]
I would recommend to train from scratch.

**Comment by saivineethkumar** [comment]
ok, thanks for your time, you helped me a lot with this.

**Comment by ahsan856jalal** [comment]
@AlexeyAB 
So I should add class names in my_dataset.shortnames.list and what exactly to add in my_dataset.labels.list file as I added same class names and when I run classifier training , I am getting too many or too few labels line shown again and again

**Comment by AlexeyAB** [comment]
@ahsan856jalal 
Try to download ImageNet dataset and create labels for Validation as said here: https://pjreddie.com/darknet/imagenet/

Download `ILSVRC2012_bbox_val_v3.tgz` and `ILSVRC2012_img_val.tar`
```
tar -xzf ILSVRC2012_bbox_val_v3.tgz
mkdir -p imgs && tar xf ILSVRC2012_img_val.tar -C imgs
```
```
wget https://pjreddie.com/media/files/imagenet_label.sh
bash imagenet_label.sh

mv inet.val.list <path-to>/darknet/data
```

So you will see what should be in labels and other files.

**Comment by zhaotun** [comment]
> @AlexeyAB
> I am able to resolve the above error and train the model. The model is predicting the new classes well but fail to retain its accuracy for the old classes ( it is not at all predicting the old classes correctly).
> 

@saivineethkumar    I got the same error, how do you resolve it ? can you tell me ? thank you so much!
